lib/elixir/lib/calendar.ex

```diff
@@ -2042,27 +2045,39 @@ defmodule DateTime do
       iex> dt = %DateTime{year: 2000, month: 2, day: 29, zone_abbr: "AMT",
       ...>                hour: 23, minute: 0, second: 7, microsecond: {0, 0},
       ...>                utc_offset: -14400, std_offset: 0, time_zone: "America/Manaus"}
-      iex> DateTime.to_iso8601(dt)
+      iex> DateTime.to_iso8601(dt, :extended)
       "2000-02-29T23:00:07-04:00"
+
+      iex> dt = %DateTime{year: 2000, month: 2, day: 29, zone_abbr: "AMT",
+      ...>                hour: 23, minute: 0, second: 7, microsecond: {0, 0},
+      ...>                utc_offset: -14400, std_offset: 0, time_zone: "America/Manaus"}
+      iex> DateTime.to_iso8601(dt, :basic)
+      "20000229T230007-0400"
   """
-  @spec to_iso8601(Calendar.datetime) :: String.t
-  def to_iso8601(datetime)
+  @spec to_iso8601(Calendar.datetime, :extended | :basic ) :: String.t
+  def to_iso8601(datetime, format \\ :extended)
 
   def to_iso8601(%{calendar: Calendar.ISO, year: year, month: month, day: day,
                   hour: hour, minute: minute, second: second, microsecond: microsecond,
-                  time_zone: time_zone, zone_abbr: zone_abbr, utc_offset: utc_offset, std_offset: std_offset}) do
+                  time_zone: time_zone, zone_abbr: zone_abbr, utc_offset: utc_offset, std_offset: std_offset}, format) when format in [:extended, :basic] do
     Calendar.ISO.datetime_to_iso8601(year, month, day, hour, minute, second, microsecond,
-                                     time_zone, zone_abbr, utc_offset, std_offset)
+                                     time_zone, zone_abbr, utc_offset, std_offset, format)
   end
 
   def to_iso8601(%{calendar: _, year: _, month: _, day: _,
                    hour: _, minute: _, second: _, microsecond: _,
-                   time_zone: _, zone_abbr: _, utc_offset: _, std_offset: _} = datetime) do
+                   time_zone: _, zone_abbr: _, utc_offset: _, std_offset: _} = datetime, format) when format in [:extended, :basic] do
     datetime
     |> convert!(Calendar.ISO)
-    |> to_iso8601
+    |> to_iso8601(format)
   end
 
+  def to_iso8601(_, format) do
+    mfa = Exception.format_mfa(__MODULE__, :to_iso8601, 2)
+    raise ArgumentError, "#{inspect format} is not a valid format argument to #{mfa}. Format must be one of `:extended` or `:basic`."
```

Comment:
No need for the `format_mfa` indirection. We also don't wrap terms in error messages in Elixir with backticks. Please write the message as:

    raise ArgumentError, "DateTime.to_iso8601/2 expects format to be :extended or :basic, got: #{inspect fotmat}"